module/zfs/dsl_scan.c

```diff
@@ -1827,6 +1827,13 @@ dsl_scan_recurse(dsl_scan_t *scn, dsl_dataset_t *ds, dmu_objset_type_t ostype,
 
 	ASSERT(!BP_IS_REDACTED(bp));
 
+	if (dnp != NULL &&
+	    dnp->dn_bonuslen > DN_MAX_BONUS_LEN(dnp)) {
+		scn->scn_phys.scn_errors++;
+		spa_log_error(dp->dp_spa, zb);
+		return (SET_ERROR(EINVAL));
+	}
```

Comment:
Detecting it and logging it seems like the most reasonable thing to do for this very unlikely case.  I think we probably could recalculate the correct size and re-dirty the dnode to fix it, but I don't think we really want to extend scrub in that way.  Let's at least add a comment mentioning how this impossible state might occur.

It'd probably make sense to update `zstreamdump` to detect this issue and at least print a warning.

Reply:
I think we are not able to correct it at all, since we don't know if it indeed has a spill block pointer.